// @mlightcad/libredwg-web/lib/database/entities/hatch.js
export var DwgHatchSolidFill;
(function (DwgHatchSolidFill) {
    DwgHatchSolidFill[DwgHatchSolidFill["PatternFill"] = 0] = "PatternFill";
    DwgHatchSolidFill[DwgHatchSolidFill["SolidFill"] = 1] = "SolidFill";
})(DwgHatchSolidFill || (DwgHatchSolidFill = {}));
export var DwgHatchAssociativity;
(function (DwgHatchAssociativity) {
    DwgHatchAssociativity[DwgHatchAssociativity["NonAssociative"] = 0] = "NonAssociative";
    DwgHatchAssociativity[DwgHatchAssociativity["Associative"] = 1] = "Associative";
})(DwgHatchAssociativity || (DwgHatchAssociativity = {}));
export var DwgHatchStyle;
(function (DwgHatchStyle) {
    DwgHatchStyle[DwgHatchStyle["Normal"] = 0] = "Normal";
    DwgHatchStyle[DwgHatchStyle["Outer"] = 1] = "Outer";
    DwgHatchStyle[DwgHatchStyle["Ignore"] = 2] = "Ignore";
})(DwgHatchStyle || (DwgHatchStyle = {}));
export var DwgHatchPatternType;
(function (DwgHatchPatternType) {
    DwgHatchPatternType[DwgHatchPatternType["UserDefined"] = 0] = "UserDefined";
    DwgHatchPatternType[DwgHatchPatternType["Predefined"] = 1] = "Predefined";
    DwgHatchPatternType[DwgHatchPatternType["Custom"] = 2] = "Custom";
})(DwgHatchPatternType || (DwgHatchPatternType = {}));
export var DwgHatchBoundaryAnnotation;
(function (DwgHatchBoundaryAnnotation) {
    DwgHatchBoundaryAnnotation[DwgHatchBoundaryAnnotation["NotAnnotated"] = 0] = "NotAnnotated";
    DwgHatchBoundaryAnnotation[DwgHatchBoundaryAnnotation["Annotated"] = 1] = "Annotated";
})(DwgHatchBoundaryAnnotation || (DwgHatchBoundaryAnnotation = {}));
export var DwgHatchGradientFlag;
(function (DwgHatchGradientFlag) {
    DwgHatchGradientFlag[DwgHatchGradientFlag["Solid"] = 0] = "Solid";
    DwgHatchGradientFlag[DwgHatchGradientFlag["Gradient"] = 1] = "Gradient";
})(DwgHatchGradientFlag || (DwgHatchGradientFlag = {}));
export var DwgHatchGradientColorFlag;
(function (DwgHatchGradientColorFlag) {
    DwgHatchGradientColorFlag[DwgHatchGradientColorFlag["TwoColor"] = 0] = "TwoColor";
    DwgHatchGradientColorFlag[DwgHatchGradientColorFlag["OneColor"] = 1] = "OneColor";
})(DwgHatchGradientColorFlag || (DwgHatchGradientColorFlag = {}));
export var DwgBoundaryPathTypeFlag;
(function (DwgBoundaryPathTypeFlag) {
    DwgBoundaryPathTypeFlag[DwgBoundaryPathTypeFlag["Default"] = 0] = "Default";
    DwgBoundaryPathTypeFlag[DwgBoundaryPathTypeFlag["External"] = 1] = "External";
    DwgBoundaryPathTypeFlag[DwgBoundaryPathTypeFlag["Polyline"] = 2] = "Polyline";
    DwgBoundaryPathTypeFlag[DwgBoundaryPathTypeFlag["Derived"] = 4] = "Derived";
    DwgBoundaryPathTypeFlag[DwgBoundaryPathTypeFlag["Textbox"] = 8] = "Textbox";
    DwgBoundaryPathTypeFlag[DwgBoundaryPathTypeFlag["Outermost"] = 16] = "Outermost";
})(DwgBoundaryPathTypeFlag || (DwgBoundaryPathTypeFlag = {}));
export var DwgBoundaryPathEdgeType;
(function (DwgBoundaryPathEdgeType) {
    DwgBoundaryPathEdgeType[DwgBoundaryPathEdgeType["Line"] = 1] = "Line";
    DwgBoundaryPathEdgeType[DwgBoundaryPathEdgeType["Circular"] = 2] = "Circular";
    DwgBoundaryPathEdgeType[DwgBoundaryPathEdgeType["Elliptic"] = 3] = "Elliptic";
    DwgBoundaryPathEdgeType[DwgBoundaryPathEdgeType["Spline"] = 4] = "Spline";
})(DwgBoundaryPathEdgeType || (DwgBoundaryPathEdgeType = {}));
