// @mlightcad/libredwg-web/lib/database/entities/polyline.js
export var DwgPolylineFlag;
(function (DwgPolylineFlag) {
    DwgPolylineFlag[DwgPolylineFlag["CLOSED_POLYLINE"] = 1] = "CLOSED_POLYLINE";
    DwgPolylineFlag[DwgPolylineFlag["CURVE_FIT"] = 2] = "CURVE_FIT";
    DwgPolylineFlag[DwgPolylineFlag["SPLINE_FIT"] = 4] = "SPLINE_FIT";
    DwgPolylineFlag[DwgPolylineFlag["POLYLINE_3D"] = 8] = "POLYLINE_3D";
    DwgPolylineFlag[DwgPolylineFlag["POLYGON_3D"] = 16] = "POLYGON_3D";
    DwgPolylineFlag[DwgPolylineFlag["CLOSED_POLYGON"] = 32] = "CLOSED_POLYGON";
    DwgPolylineFlag[DwgPolylineFlag["POLYFACE"] = 64] = "POLYFACE";
    DwgPolylineFlag[DwgPolylineFlag["CONTINUOUS"] = 128] = "CONTINUOUS";
})(DwgPolylineFlag || (DwgPolylineFlag = {}));
export var DwgSmoothType;
(function (DwgSmoothType) {
    DwgSmoothType[DwgSmoothType["NONE"] = 0] = "NONE";
    DwgSmoothType[DwgSmoothType["QUADRATIC"] = 5] = "QUADRATIC";
    DwgSmoothType[DwgSmoothType["CUBIC"] = 6] = "CUBIC";
    DwgSmoothType[DwgSmoothType["BEZIER"] = 8] = "BEZIER";
})(DwgSmoothType || (DwgSmoothType = {}));
