// @mlightcad/libredwg-web/lib/database/entities/text.js
export var DwgTextGenerationFlag;
(function (DwgTextGenerationFlag) {
    DwgTextGenerationFlag[DwgTextGenerationFlag["NONE"] = 0] = "NONE";
    DwgTextGenerationFlag[DwgTextGenerationFlag["MIRRORED_X"] = 2] = "MIRRORED_X";
    DwgTextGenerationFlag[DwgTextGenerationFlag["MIRRORED_Y"] = 4] = "MIRRORED_Y";
})(DwgTextGenerationFlag || (DwgTextGenerationFlag = {}));
export var DwgTextHorizontalAlign;
(function (DwgTextHorizontalAlign) {
    DwgTextHorizontalAlign[DwgTextHorizontalAlign["LEFT"] = 0] = "LEFT";
    DwgTextHorizontalAlign[DwgTextHorizontalAlign["CENTER"] = 1] = "CENTER";
    DwgTextHorizontalAlign[DwgTextHorizontalAlign["RIGHT"] = 2] = "RIGHT";
    DwgTextHorizontalAlign[DwgTextHorizontalAlign["ALIGNED"] = 3] = "ALIGNED";
    DwgTextHorizontalAlign[DwgTextHorizontalAlign["MIDDLE"] = 4] = "MIDDLE";
    DwgTextHorizontalAlign[DwgTextHorizontalAlign["FIT"] = 5] = "FIT";
})(DwgTextHorizontalAlign || (DwgTextHorizontalAlign = {}));
export var DwgTextVerticalAlign;
(function (DwgTextVerticalAlign) {
    DwgTextVerticalAlign[DwgTextVerticalAlign["BASELINE"] = 0] = "BASELINE";
    DwgTextVerticalAlign[DwgTextVerticalAlign["BOTTOM"] = 1] = "BOTTOM";
    DwgTextVerticalAlign[DwgTextVerticalAlign["MIDDLE"] = 2] = "MIDDLE";
    DwgTextVerticalAlign[DwgTextVerticalAlign["TOP"] = 3] = "TOP";
})(DwgTextVerticalAlign || (DwgTextVerticalAlign = {}));
